pkg/workflows/exec/interpolation.go

```diff
@@ -0,0 +1,100 @@
+package exec
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/values"
+	"github.com/smartcontractkit/chainlink-common/pkg/workflows"
+)
+
+// InterpolateKey takes a multi-part, dot-separated key and attempts to replace
+// it with its corresponding value in `state`.
+//
+// A key is valid if it contains at least two parts, with:
+//   - the first part being the workflow step's `ref` variable
+//   - the second part being one of `inputs` or `outputs`
+//
+// If a key has more than two parts, then we traverse the parts
+// to find the value we want to replace.
+// We support traversing both nested maps and lists and any combination of the two.
+func InterpolateKey(key string, state Results) (any, error) {
```

Comment:
Do we need to use an interface here? Under the hood this is just a map of State objects isn't it? Why not move that out too -- that feels like it simplifies the implementation

Reply:
If that is not possible, another in-between approach could be to accept a func here:
```suggestion
func InterpolateKey(key string, resultForStep func(string) (*Result, bool)) (any, error) {
```
Then caller just passes a reference to its method: `exec.InterpolateKey(key, results.ResultsForStep)